update docs to use security policy

mmerickel opened the pull request:
- [X] rename `identify` to `authenticated_identity` on `ISecurityPolicy`
- [X] quick_tutorial
- [x] sync cookiecutter to wiki tutorial
- [x] use security policy in wiki tutorial
- [X] sync cookiecutter to wiki2
- [X] use security policy in wiki2 tutorial
- [x] revise security chapter - this is at least partially complete
- [x] sync cookiecutter to starter project

rendered: https://mmerickel-pyramid-fork.readthedocs.io/en/security-docs/

fixes #3548

stevepiercy commented on docs/quick_tutorial/authorization.rst:
```diff
@@ -87,8 +92,10 @@ This simple tutorial step can be boiled down to the following:
 
-- The registered ``groupfinder`` answers whether a particular user (``editor``)
-  has a particular group (``group:editors``).
+- The ``SecurityPolicy.effective_principals`` method answers whether a particular user (``editor``) has a particular group (``group:editors``).
```
```suggestion
- The ``SecurityPolicy.effective_principals`` method answers whether a particular user (``editor``) belongs to a particular group (``group:editors``).
```

stevepiercy replied:
Alternatively, "is a member of".

mmerickel commented:
Please do sync the cookiecutters. Waiting on that to finish this PR.

stevepiercy commented:
Is your open PR finished? https://github.com/Pylons/pyramid-cookiecutter-starter/pull/81

mmerickel commented:
yes

mmerickel commented:
Not sure why you approved this - it isn't done yet.

stevepiercy commented:
I approved work thusfar. I can do another one later, too. It sounds like you prefer one final review going forward, which I will do.

mmerickel commented:
It’s fine I guess. The PR is in draft mode meaning I don’t really care if anyone reviews it yet.

mmerickel commented:
I can sync the cookiecutters if you don’t have time. Just gotta let me know.

stevepiercy commented:
I'm working on it.

mmerickel commented:
Ok!

stevepiercy commented:
This PR depends on the completion of #3556

merwok commented on docs/quick_tutorial/authentication.rst:
```diff
@@ -111,14 +110,12 @@ are you) and authorization (what are you allowed to do) are not just pluggable,
 but decoupled. To learn one step at a time, we provide a system that identifies
```
Aren’t they coupled in SecurityPolicy now?

mmerickel replied:
clarified in ef5b40196

mmerickel commented:
Ok, I've significantly updated the wiki2 tutorial.

1. Synced it with the starter cookiecutter.
1. Added CSRF protection to the app in the view chapter.
1. Added a ``MySecurityPolicy`` in authentication and authorization chapters.
1. Significant revamping of the test chapter, including rewriting everything to use pytest, with useful transactional fixtures.
1. New release of pyramid_tm (2.4) to support the new fixtures and to stop using `request.unauthenticated_userid`.
1. New release of pyramid_debugtoolbar (4.5.2) to stop using `request.unauthenticated_userid`.

I still need to sync the cookiecutter with wiki tutorial (I won't be nearly as detailed in my changes there unfortunately), and to review the security chapter documentation. There may also be some missing cookiecutter udpates in the starter project chapter.

merwok commented on docs/tutorials/wiki2/authentication.rst:
```diff
@@ -215,6 +197,9 @@ This code adds three new views to the application:
   The check is done by first finding a ``User`` record in the database, then
   using our ``user.check_password`` method to compare the hashed passwords.
 
+  At a privilege boundary we are sure to reset the CSRF token using :meth:`pyramid.csrf.new_csrf_token`.
```
Could you add a line to explain why this is useful?

mmerickel replied:
I would prefer to just provide the pattern. I think it's distracting to go into too many details in the tutorial. I'd accept something if you added one, especially to the CSRF protection docs in the security chapter - but ideally not here.

merwok replied:
I couldn’t add something, as I don’t understand the reason now :slightly_smiling_face: 

Thanks for showing the pattern nonetheless!  I don’t see harm in copying it without full understanding, and I may understand the reasons after I read some articles in my reading list.

mmerickel replied:
It's definitely not as important for a CSRF token - which mostly needs to be unique per client so that an attacking website cannot guess it. It's way more important for session data. The general recommendation is to always clear application state when someone logs in or logs out. You don't want to accidentally access random state with a logged-in user that was set by someone else before the user logged in. It can trigger unintended side-effects and unanticipated behaviors.

merwok replied:
Thanks for explaining.  I haven’t encountered these issues before, as I work on apps that don’t use sessions (clients drive interactions, or if state has to be persistent then it’s application state that is saved in database).  I guess I haven’t had any fun use case yet like «items added to cart before login should be kept after»!

merwok commented on docs/tutorials/wiki2/src/tests/tutorial/security.py:
```diff
@@ -8,33 +9,50 @@
+
+    config.set_security_policy(MySecurityPolicy(settings['auth.secret']))
+    config.add_request_method(
+        lambda request: request.authenticated_identity, 'user', property=True)
```
Why not `reify` to add caching?

mmerickel replied:
`request.user` is not the source of truth. Caching is hard - if you cache something that is not the source, you're doing it wrong. Even if I rewrote the policy to use `request.user` everywhere, `request.authenticated_identity` is still an api, and would not be cached. Caching needs to be done in `authenticated_identity` by the policy, if we're going to do it. In this case it is already basically cached because there are no database queries on subsequent calls thanks to SQLAlchemy's identity map.

merwok commented:
Lots of good stuff here.  Thank you for this work!

stevepiercy commented:
> I still need to sync the cookiecutter with wiki tutorial (I won't be nearly as detailed in my changes there unfortunately), and to review the security chapter documentation. There may also be some missing cookiecutter udpates in the starter project chapter.

I can do the grunt work here to sync up details. Thank you for doing all the stuff that requires thought, knowledge, and skill.

mmerickel commented:
@stevepiercy if you want to sync the cookiecutter to the starter project please be my guest. I'll do the wiki tutorial over the next few days.

stevepiercy commented:
@mmerickel just to make sure I understand correctly, "sync cookiecutter to starter project" == docs that refer to source files in `docs/narr/myproject`?

mmerickel commented:
@stevepiercy yes that’s accurate

stevepiercy commented:
@mmerickel https://github.com/mmerickel/pyramid/pull/10 submitted for your review to handle sync cookiecutter to starter project item.

mmerickel commented:
great thank you @stevepiercy

stevepiercy commented on docs/tutorials/wiki2/authentication.rst:
```diff
@@ -51,15 +51,30 @@ It also handles authorization, which we'll cover in the next chapter (if you're
+- It provides consistency across method invocations to ensure the identity does not change while processing the request.
+
+It is up to individual security policies and applications to determine the best approach with respect to caching.
+Applications is long-running requests may want to avoid caching the identity, or tracking some extra metadata to re-verify it periodically against the authentication source.
```
"Applications with"?

mmerickel commented:
I've got 2 chapters left in the wiki tutorial then I think we're done with this.....

mmerickel commented:
ok this is ready for review, I added a rendered version at https://mmerickel-pyramid-fork.readthedocs.io/en/security-docs/ to view the wiki and wiki2 tutorials more easily since it's so damn many changes.

stevepiercy commented:
@mmerickel against which branch of Pyramid did you run tests for the wiki2 tutorial? I tried both:

```
pip install git+https://github.com/Pylons/pyramid@master
pip install git+https://github.com/mmerickel/pyramid@security-docs
```

And I'm getting `13 failed, 14 passed in 9.49s`.

mmerickel commented:
You need to install pyramid master into the virtualenv you create for the tutorial step. After that, you need to post an actual error message.

mmerickel commented on docs/tutorials/wiki2/tests.rst:
```diff
@@ -88,7 +88,7 @@ Update ``tests/conftest.py`` to look like the following, adding the highlighted
 .. literalinclude:: src/tests/tests/conftest.py
     :linenos:
-    :emphasize-lines: 10,68-103,110,117-119
+    :emphasize-lines: 68-103,110,117-119
```
Revert this, the import wasn't supposed to be in the cookiecutter.

stevepiercy replied:
Done in 6d3e713

mmerickel commented on docs/tutorials/wiki/installation.rst:
```diff
@@ -317,7 +317,7 @@ On Unix
 .. code-block:: bash
 
-    $VENV/bin/pytest --cov=tutorial tests.py -q
+    $VENV/bin/pytest --cov=tutorial tests -q
```
your invocations of pytest are inconsistent across chapters. The folder does not need to be specified and notice above on line 264/272 you used `--cov` not `--cov=tutorial`.

stevepiercy replied:
Check the context for this invocation:

> Without these defaults, we would need to specify the path to the module on which we want to run tests and coverage.

To verify, I moved both the default `pytest.ini` and `.coveragerc` out of the project directory.

```
$VENV/bin/pytest --cov=tutorial -q
11 errors in 5.41s

$VENV/bin/pytest --cov=tutorial tests -q
# succeeds
```

Invocations of pytest in `wiki/installation.rst` were copied from `wiki2/installation.rst` verbatim.

I did find a couple of other unrelated things that needed to be updated, though, and pushed those commits.

mmerickel replied:
we have the testpaths set in pytest.ini, you do not need to specify the tests folder.

stevepiercy replied:
Then we need to remove the context as well from both tutorials because it would repeat an earlier step.

https://github.com/Pylons/pyramid/pull/3557/files/56fd7cbeaf107405cca2399d2b0bba60c8e60010..04cef9070f8554a9a025ae968368d03b4ed939c2#diff-11018ed84544958d5baea4415d2639edR307-R320

I think it is useful to retain this step because it demonstrates the verbosity of commands when you do not have the default configuration files. Your call.

mmerickel replied:
well the cookiecutter output just says to run `pytest` and the point of that config was to keep the calling convention simple so I'd rather it minimize the number of cli args being passed

stevepiercy replied:
Yes, I understand the purpose of adding the config updates.

I am asking whether you want to remove these two sections, which explain what one needs to do without those two config files (and implies why config files are a good thing™), as well as links to more pytest options:

https://github.com/Pylons/pyramid/pull/3557/files/56fd7cbeaf107405cca2399d2b0bba60c8e60010..04cef9070f8554a9a025ae968368d03b4ed939c2#diff-11018ed84544958d5baea4415d2639edR305-R335

https://github.com/Pylons/pyramid/pull/3557/files#diff-e0805d6d677c4e8151d97c2a617fe1dcR384-R416

mmerickel replied:
Yes I think I'd prefer to remove the section saying "if things weren't this way then you'd have to do this" because they can just appreciate our config and read the pytest docs themselves. I'd keep the part mentioning that the config files exist.

stevepiercy replied:
OK, please check out commit 3152aa5.

stevepiercy commented:
> You need to install pyramid master into the virtualenv you create for the tutorial step. After that, you need to post an actual error message.

`$VENV/bin/pip install git+https://github.com/Pylons/pyramid@master`

Ensured that the source files in `docs/tutorials/wiki2/src/tests` match those in my project.

`$VENV/bin/pytest -q`

```
# Just the last test failure:

__________________________________________________________________________________________________________________________________________ test_editors_member_user_can_view ___________________________________________________________________________________________________________________________________________

request = <Request at 0x108d35240 GET http://example.com/FrontPage>, exc = UndefinedError("'pyramid.request.Request object' has no attribute 'user'")

    def _error_handler(request, exc):
        # NOTE: we do not need to delete exc_info because this function
        # should never be in the call stack of the exception
        exc_info = sys.exc_info()
    
        try:
>           response = request.invoke_exception_view(exc_info)

lib/python3.7/site-packages/pyramid/tweens.py:13: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Request at 0x108d35240 GET http://example.com/FrontPage>, exc_info = (<class 'jinja2.exceptions.UndefinedError'>, UndefinedError("'pyramid.request.Request object' has no attribute 'user'"), <traceback object at 0x10921b8c8>), request = <Request at 0x108d35240 GET http://example.com/FrontPage>
secure = True, reraise = False

    def invoke_exception_view(
        self, exc_info=None, request=None, secure=True, reraise=False
    ):
        """ Executes an exception view related to the request it's called upon.
        The arguments it takes are these:
    
        ``exc_info``
    
            If provided, should be a 3-tuple in the form provided by
            ``sys.exc_info()``.  If not provided,
            ``sys.exc_info()`` will be called to obtain the current
            interpreter exception information.  Default: ``None``.
    
        ``request``
    
            If the request to be used is not the same one as the instance that
            this method is called upon, it may be passed here.  Default:
            ``None``.
    
        ``secure``
    
            If the exception view should not be rendered if the current user
            does not have the appropriate permission, this should be ``True``.
            Default: ``True``.
    
        ``reraise``
    
            A boolean indicating whether the original error should be reraised
            if a :term:`response` object could not be created. If ``False``
            then an :class:`pyramid.httpexceptions.HTTPNotFound`` exception
            will be raised. Default: ``False``.
    
        If a response is generated then ``request.exception`` and
        ``request.exc_info`` will be left at the values used to render the
        response. Otherwise the previous values for ``request.exception`` and
        ``request.exc_info`` will be restored.
    
        .. versionadded:: 1.7
    
        .. versionchanged:: 1.9
           The ``request.exception`` and ``request.exc_info`` properties will
           reflect the exception used to render the response where previously
           they were reset to the values prior to invoking the method.
    
           Also added the ``reraise`` argument.
    
        """
        if request is None:
            request = self
        registry = getattr(request, 'registry', None)
        if registry is None:
            registry = get_current_registry()
    
        if registry is None:
            raise RuntimeError("Unable to retrieve registry")
    
        if exc_info is None:
            exc_info = sys.exc_info()
    
        exc = exc_info[1]
        attrs = request.__dict__
        context_iface = providedBy(exc)
    
        # clear old generated request.response, if any; it may
        # have been mutated by the view, and its state is not
        # sane (e.g. caching headers)
        with hide_attrs(request, 'response', 'exc_info', 'exception'):
            attrs['exception'] = exc
            attrs['exc_info'] = exc_info
            # we use .get instead of .__getitem__ below due to
            # https://github.com/Pylons/pyramid/issues/700
            request_iface = attrs.get('request_iface', IRequest)
    
            manager.push({'request': request, 'registry': registry})
    
            try:
                response = _call_view(
                    registry,
                    request,
                    exc,
                    context_iface,
                    '',
                    view_types=None,
                    view_classifier=IExceptionViewClassifier,
                    secure=secure,
                    request_iface=request_iface.combined,
                )
            except Exception:
                if reraise:
                    reraise_(*exc_info)
                raise
            finally:
                manager.pop()
    
        if response is None:
            if reraise:
                reraise_(*exc_info)
>           raise HTTPNotFound
E           pyramid.httpexceptions.HTTPNotFound: The resource could not be found.

lib/python3.7/site-packages/pyramid/view.py:770: HTTPNotFound

During handling of the above exception, another exception occurred:

testapp = <tests.conftest.TestApp object at 0x108d35438>

    def test_editors_member_user_can_view(testapp):
        testapp.login(editor_login)
>       res = testapp.get('/FrontPage', status=200)

tests/test_functional.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/python3.7/site-packages/webtest/app.py:335: in get
    expect_errors=expect_errors)
lib/python3.7/site-packages/webtest/app.py:628: in do_request
    res = req.get_response(app, catch_exc_info=True)
lib/python3.7/site-packages/webob/request.py:1310: in send
    application, catch_exc_info=True)
lib/python3.7/site-packages/webob/request.py:1278: in call_application
    app_iter = application(self.environ, start_response)
lib/python3.7/site-packages/webtest/lint.py:201: in lint_app
    iterator = application(environ, start_response_wrapper)
lib/python3.7/site-packages/pyramid/router.py:266: in __call__
    response = self.execution_policy(environ, self)
lib/python3.7/site-packages/pyramid_retry/__init__.py:121: in retry_policy
    response = router.invoke_request(request)
lib/python3.7/site-packages/pyramid/router.py:245: in invoke_request
    response = handle_request(request)
lib/python3.7/site-packages/pyramid_tm/__init__.py:120: in tm_tween
    return handler(request)
lib/python3.7/site-packages/pyramid/tweens.py:43: in excview_tween
    response = _error_handler(request, exc)
lib/python3.7/site-packages/pyramid/tweens.py:17: in _error_handler
    reraise(*exc_info)
lib/python3.7/site-packages/pyramid/util.py:740: in reraise
    raise value
lib/python3.7/site-packages/pyramid/tweens.py:41: in excview_tween
    response = handler(request)
lib/python3.7/site-packages/pyramid/router.py:144: in handle_request
    registry, request, context, context_iface, view_name
lib/python3.7/site-packages/pyramid/view.py:658: in _call_view
    response = view_callable(context, request)
lib/python3.7/site-packages/pyramid/viewderivers.py:321: in secured_view
    return view(context, request)
lib/python3.7/site-packages/pyramid/viewderivers.py:514: in csrf_view
    return view(context, request)
lib/python3.7/site-packages/pyramid/viewderivers.py:453: in rendered_view
    request, result, view_inst, context
lib/python3.7/site-packages/pyramid/renderers.py:443: in render_view
    return self.render_to_response(response, system, request=request)
lib/python3.7/site-packages/pyramid/renderers.py:466: in render_to_response
    result = self.render(value, system_values, request=request)
lib/python3.7/site-packages/pyramid/renderers.py:462: in render
    result = renderer(value, system_values)
lib/python3.7/site-packages/pyramid_jinja2/__init__.py:265: in __call__
    return template.render(system)
lib/python3.7/site-packages/jinja2/asyncsupport.py:76: in render
    return original_render(self, *args, **kwargs)
lib/python3.7/site-packages/jinja2/environment.py:1008: in render
    return self.environment.handle_exception(exc_info, True)
lib/python3.7/site-packages/jinja2/environment.py:780: in handle_exception
    reraise(exc_type, exc_value, tb)
lib/python3.7/site-packages/jinja2/_compat.py:37: in reraise
    raise value.with_traceback(tb)
tutorial/templates/view.jinja2:1: in top-level template code
    {% extends 'layout.jinja2' %}
lib/python3.7/site-packages/jinja2/environment.py:1005: in render
    return concat(self.root_render_func(self.new_context(vars)))
tutorial/templates/view.jinja2:14: in root
    </p>
tutorial/templates/layout.jinja2:31: in root
    <div class="col-md-2">
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyramid_jinja2.Environment object at 0x1088f4898>, obj = Undefined, attribute = 'name'

    def getattr(self, obj, attribute):
        """Get an item or attribute of an object but prefer the attribute.
        Unlike :meth:`getitem` the attribute *must* be a bytestring.
        """
        try:
>           return getattr(obj, attribute)
E           jinja2.exceptions.UndefinedError: 'pyramid.request.Request object' has no attribute 'user'

lib/python3.7/site-packages/jinja2/environment.py:430: UndefinedError
13 failed, 14 passed in 9.52s
```

mmerickel commented:
You need to install this branch, not master.

mmerickel commented:
I do it with `pip install -e ~/work/oss/pyramid` myself.

merwok commented on docs/quick_tutorial/authorization.rst:
```diff
@@ -87,8 +92,10 @@ This simple tutorial step can be boiled down to the following:
 
-- The registered ``groupfinder`` answers whether a particular user (``editor``)
-  has a particular group (``group:editors``).
+- The ``SecurityPolicy.effective_principals`` method answers whether a particular user (``editor``) is a member of a particular group (``group:editors``).
```
Would it be bad to inline the code from `effective_principals` inside the `permits` method?
It is not an interface method anymore, just an internal helper for this specific security policy (similar to `load_identity` with RequestLocalCache); I feel that not having a function named `effective_principals` here would help avoiding confusion now that the whole concept is deprecated (at a framework level anyway, a project could still use it).

mmerickel replied:
I thought about this but I think I'd prefer to leave it as is as the code is cleaner and it's only an issue for someone like you who already knows how Pyramid's system works. Also I don't want to stop people from capturing a ref to the policy somewhere and invoking other methods on it like this one if they need to.

mmerickel replied:
The only change I'd consider here is pushing the principals into the authenticated_identity and pulling them out of there. This would require constructing a custom identity object instead of using the `User`. For example something like a dict with `user` and `effective_principals` keys. At least then `request.user` wouldn't just be an alias for the identity if I did that.

I'm still leaning toward keeping it the way it is though.

merwok replied:
Using dicts or a named tuple for users / identity objects make sense to me!

mmerickel replied:
Sorry I thought about it again and am gonna leave it as is.

stevepiercy commented:
> You need to install this branch, not master.

Yup, did that as well, both now and earlier, unless this is not the correct command to install this branch.

`pip install git+https://github.com/mmerickel/pyramid@security-docs`

Still got the exact same error message as above.

stevepiercy commented:
I tossed out my virtual environment, rebuilt the tutorial from scratch, and now all tests pass.

`27 passed in 6.91s`

mmerickel commented:
Thanks @stevepiercy I think this PR is ready to be merged.

stevepiercy commented:
Not yet. I still have the last two steps to do in the traversal wiki. Will do late tonight.

stevepiercy commented:
OK, I'm done! Any final revisions?

mmerickel commented:
Thanks @stevepiercy and @merwok for your review!
